datafusion/physical-plan/src/aggregates/group_values/row_like.rs

```diff
@@ -0,0 +1,392 @@
+                        &DataType::UInt64 => {
+                            let b = PrimitiveGroupValueBuilder::<UInt64Type>::new();
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::Float32 => {
+                            let b = PrimitiveGroupValueBuilder::<Float32Type>::new();
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::Float64 => {
+                            let b = PrimitiveGroupValueBuilder::<Float64Type>::new();
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::Utf8 => {
+                            let b =
+                                ByteGroupValueBuilderNaive::<i32>::new(OutputType::Utf8);
+                            v.push(Box::new(b) as _)
+                        }
+                        &DataType::LargeUtf8 => {
+                            let b =
+                                ByteGroupValueBuilderNaive::<i64>::new(OutputType::Utf8);
+                            v.push(Box::new(b) as _)
+                        }
+                        dt => todo!("{dt} not impl"),
+                    }
+                }
+                v
+            }
+        };
+
+        // tracks to which group each of the input rows belongs
+        groups.clear();
+
+        // 1.1 Calculate the group keys for the group values
+        let batch_hashes = &mut self.hashes_buffer;
+        batch_hashes.clear();
+        batch_hashes.resize(n_rows, 0);
+        create_hashes(cols, &self.random_state, batch_hashes)?;
+
+        for (row, &target_hash) in batch_hashes.iter().enumerate() {
+            let entry = self.map.get_mut(target_hash, |(exist_hash, group_idx)| {
+                // Somewhat surprisingly, this closure can be called even if the
+                // hash doesn't match, so check the hash first with an integer
+                // comparison first avoid the more expensive comparison with
+                // group value. https://github.com/apache/datafusion/pull/11718
+                if target_hash != *exist_hash {
+                    return false;
+                }
+                // verify that the group that we are inserting with hash is
+                // actually the same key value as the group in
+                // existing_idx  (aka group_values @ row)
+                // && group_rows.row(row) == group_values.row(*group_idx)
+
+                fn compare_equal(
+                    arry_eq: &dyn ArrayEq,
+                    lhs_row: usize,
+                    array: &ArrayRef,
+                    rhs_row: usize,
+                ) -> bool {
+                    arry_eq.equal_to(lhs_row, array, rhs_row)
+                }
+
+                for (i, group_val) in group_values_v2.iter().enumerate() {
+                    if !compare_equal(group_val.as_ref(), *group_idx, &cols[i], row) {
```

Comment:
As this is called in a loop, this can be optimized/specialized for certain cases like: do the arrays have any nulls or not.

Reply:
I don't get it how could I further optimize the loop based on nulls 🤔

Reply:
Yes, indeed 👍